### PR TITLE
TreeView - fix selection changed event raised twice (T988753)

### PR DESCRIPTION
**Pull request by novsstation**
Bugfix T988753

### CODE REVIEW

File: js/ui/tree_view/ui.tree_view.base.js [hunk 1]
```diff
@@ -1217,7 +1226,9 @@ const TreeViewBase = HierarchicalCollectionWidget.inherit({
         });
 
         if(this._selectAllEnabled()) {
-            this._$selectAllItem.dxCheckBox('instance').option('value', this._dataAdapter.isAllSelected());
```
Comment (novsstation) [inline js/ui/tree_view/ui.tree_view.base.js:1220]: _updateItemsUI must only update UI. It should not fire events.

So I rewrite it with _setOptionWithoutOptionChange and repaint

File: js/ui/tree_view/ui.tree_view.base.js [hunk 2]
```diff
@@ -1217,7 +1226,9 @@ const TreeViewBase = HierarchicalCollectionWidget.inherit({
...
+            const selectAllCheckbox = this._getSelectAllCheckbox();
+            selectAllCheckbox._setOptionWithoutOptionChange('value', this._dataAdapter.isAllSelected());
+            selectAllCheckbox.repaint();
```
Comment (IgnatovDan) [inline js/ui/tree_view/ui.tree_view.base.js]: Similar to React: rerender component if its state was changed.

File: testing/tests/DevExpress.ui.widgets/treeViewParts/checkboxes.js [hunk 3]
```diff
@@ -213,6 +213,58 @@ QUnit.test('T195986', function(assert) {
...
+        });
+    });
+});
+
+QUnit.test('Check value of the selectAllValueChanged event (T988753)', function(assert) {
+    let eventValue;
+    const wrapper = new TreeViewTestWrapper({
+        showCheckBoxesMode: 'selectAll',
+        items: [ { text: 'item1', expanded: true, items: [ { text: 'item1_1' }, { text: 'item1_2' } ] } ],
+        onSelectAllValueChanged: (args) => { eventValue = args.value; }
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.ui.widgets/treeViewParts/checkboxes.js]: Whether 'selectAllValueChangedArrayLog' will be better to track event calls?

File: testing/tests/DevExpress.ui.widgets/treeViewParts/checkboxes.js [hunk 4]
```diff
@@ -213,6 +213,58 @@ QUnit.test('T195986', function(assert) {
     assert.strictEqual($(checkboxes[0]).dxCheckBox('instance').option('value'), true);
 });
 
+['none', 'normal', 'selectAll'].forEach((showCheckBoxesMode) => {
+    ['multiple', 'single'].forEach((selectionMode) => {
+        [false, true].forEach((selectNodesRecursive) => {
+            QUnit.test(`Click by checkbox. checkboxMode: ${showCheckBoxesMode}, selectionMode: ${selectionMode}, recursive: ${selectNodesRecursive} (T988753)`, function() {
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.ui.widgets/treeViewParts/checkboxes.js]: "All selected -> deselect middle level item" - rename?

**Review by IgnatovDan — CHANGES_REQUESTED** [review]

File: js/ui/tree_view/ui.tree_view.base.js [hunk 5]
```diff
@@ -310,6 +310,10 @@ const TreeViewBase = HierarchicalCollectionWidget.inherit({
         }
     },
 
+    _getSelectAllCheckbox: function() {
```
Comment (IgnatovDan) [inline js/ui/tree_view/ui.tree_view.base.js]: imho it doesn't worth it to add such little methods

**Review by IgnatovDan — CHANGES_REQUESTED** [review]

File: testing/tests/DevExpress.ui.widgets/treeViewParts/checkboxes.js [hunk 6]
```diff
@@ -213,6 +213,87 @@ QUnit.test('T195986', function(assert) {
     assert.strictEqual($(checkboxes[0]).dxCheckBox('instance').option('value'), true);
 });
 
+['none', 'normal', 'selectAll'].forEach((showCheckBoxesMode) => {
+    ['multiple', 'single'].forEach((selectionMode) => {
+        [false, true].forEach((selectNodesRecursive) => {
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.ui.widgets/treeViewParts/checkboxes.js:218]: Also, there is the https://js.devexpress.com/Documentation/21_1/ApiReference/UI_Components/dxTreeView/Configuration/#selectByClick option but I am not sure that we have to add tests for the same selection events in this PR

**Review by IgnatovDan — APPROVED** [review]

**Pull request merged**